## Gather

In [1]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name            surname  \
313         314       female   Mgbankwo          Afamefuna   
112         113       female  Arnþrúður  Skarphéðinsdóttir   
3             4         male       Liêm               Phan   
369         370       female    Farizah            Sleiman   
67           68         male    Nebechi          Ekechukwu   

                 address          city state  zip_code        country  \
313   1774 George Avenue        Atmore    AL   36502.0  United States   
112         2578 Tenmile  Newport News    VA   23602.0  United States   
3    2335 Webster Street    Woodbridge    NJ    7095.0  United States   
369     2322 Grim Avenue     San Diego    CA   92101.0  United States   
67     2418 Smith Street      Marlboro    MA    1752.0  United States   

                                               contact  birthdate  weight  \
313           251-359-2088MgbankwoAfamefuna@einrot.com  7/18/1984   188.5   
112  ArnruurSkarpheinsdottir@rhyta.com+1 (757) 885-...   6/4/1922   172.0   
3           PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9   
369       619-652-0285FarizahAminehSleiman@teleworm.us  8/19/1943   190.5   
67            NebechiEkechukwu@teleworm.us508-804-4850  1/11/1945   154.9   

     height   bmi  
313      67  29.5  
112      63  30.5  
3        70  31.7  
369      62  34.8  
67       64  26.6

In [15]:
patients.surname.value_counts()

Doe           6
Taylor        3
Jakobsen      3
Aranda        2
Lund          2
             ..
Horvat        1
Fiorentino    1
Totth         1
Cotton        1
Knutsen       1
Name: surname, Length: 466, dtype: int64

In [16]:
patients.address.value_counts()

123 Main Street          6
2778 North Avenue        2
648 Old Dear Lane        2
2476 Fulton Street       2
3145 Sheila Lane         1
                        ..
3251 Radio Park Drive    1
3209 Crowfield Road      1
1790 Nutter Street       1
2945 Ferguson Street     1
4853 University Drive    1
Name: address, Length: 483, dtype: int64

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality


##### `patients` table
- Zip code is a `float` not a `string`
- Zip code has four digits sometimes
- Tim `Neudorf height` is 27 in instead of 72 in
- Full `state` names sometimes, abbreviations other times
- `Dsvid` Gustafsson should be David id number 9
- Missing demographic information (`address` - `contact` columns) ***(can't clean yet)***
- Erroneous datatypes (assigned `sex`, `state`, `zip_code`, and `birthdate` columns)
- Multiple `phone number` formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva `weight`


##### `treatments` table
- Missing `HbA1c` changes
- The letter 'u' in starting and ending doses for `Auralin` and `Novodra`
- Lowercase `given names` and `surnames`
- Missing records (280 instead of 350)
- Erroneous datatypes (`auralin` and `novodra` columns)
- Inaccurate `HbA1c` changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in `auralin` and `novodra` columns


##### `adverse_reactions` table
- Lowercase `given names` and `surnames`

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

#### Notes: 
- Date of birth as month/day/year instead of day/month/year isn't a data quality issue because either representation is fine and the entire birthdate column consists of one format.

- Accents on given names and surnames (e.g., Tám Liễu) aren't data quality issues because this is how these names are actually spelled. You will have to be aware of these accents whenever performing operations that require this text (merging tables on the name column, for example), but they aren't quality issues.

- Weight as a float type (one decimal) and height as an integer (no decimals) is fine. This is just how the data were recorded. Though height may therefore be technically slightly inaccurate (by portions of an inch), this is an example of an acceptable limitation.

## Clean

In [23]:
# Always make copies of the original pieces of data before cleaning!
patients_clean = patients.copy()

In [24]:
treatments_clean = treatments.copy()

In [25]:
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Define
Import the cut treatments into a DataFrame and concatenate it with the original treatments
DataFrame.

##### Code

In [26]:
# Your cleaning code here

# Import the cut treatments into a DataFrame
treatments_cut = pd.read_csv('treatments_cut.csv')

# Concatenate it with the original treatments DataFrame.
treatments_clean = pd.concat([treatments, treatments_cut], axis='rows', ignore_index=True)

##### Test

In [27]:
# Your testing code here
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


In [28]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)


##### Define
Recalculate the hba1c_change column: hba1c_start minus hba1c_end.

##### Code

In [29]:
# Your cleaning code here
treatments_clean.hba1c_change = treatments_clean.hba1c_start - treatments_clean.hba1c_end

##### Test

In [30]:
# Your testing code here
treatments_clean.hba1c_change.head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

In [31]:
treatments_clean.hba1c_change.isnull().sum()

0

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
Extract the phone number and email variables from the contact column using regular
expressions and pandas’ str.extract method. Drop the contact column when done

##### Code

In [32]:
patients_clean.contact.sample(3)

367              775-358-9076AvaFisher@rhyta.com
17      RoxanneAndreyeva@armyspy.com678-829-8578
253    321-287-0484MagyarKrisztina@superrito.com
Name: contact, dtype: object

In [33]:
# Your cleaning code here

# Extract the phone number and email variables from the contact column using regular expressions
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', 
                                                                    expand=True)
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z_]+@[a-zA-Z]+\.[a-zA-Z]+)', expand=True)

# Drop the contact column when done
patients_clean.drop('contact', axis=1, inplace=True)

##### Test

In [34]:
# Confirm contact column is gone
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
 13  phone_number  491 non-null    object 
 14  email         491 non-null    object 
dtypes: float64(3), int64(2), object(10)
memory usage: 59.1+ KB


In [35]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sample(6)

361    HansPeterRadislav@gustr.com
89             KwasiMensa@cuvox.de
426     RogelioJTaylor@teleworm.us
437      AlwinSvensson@armyspy.com
279        MariFilatov@fleckens.hu
310           HugoCollins@cuvox.de
Name: email, dtype: object

In [36]:
# Confirm that no phone_number start with an string (regex didn't match for this)
patients_clean.phone_number.sample(6)

250         410-284-8935
93          786-234-0038
424    +1 (401) 419-2975
195         985-253-7891
75     +1 (510) 498-6624
377         217-485-5673
Name: phone_number, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain start_dose and end_dose columns. Drop the intermediate dose column.

##### Code

In [37]:
# Your cleaning code here

# Melt the auralin and novodra columns
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
        var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]

# split the dose column on ' - ' to obtain start_dose and end_dose columns.
treatments_clean['start_dose'], treatments_clean['end_dose'] = treatments_clean.dose.str.split('-').str

# Drop the intermediate dose column.
treatments_clean.drop('dose', axis=1, inplace=True)

C:\Users\Mahmoud\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  if __name__ == '__main__':


##### Test

In [38]:
# Your testing code here
treatments_clean.sample(6)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
196      peter    pospíšil         7.76       7.34          0.42   auralin   
575       liêm        phan         7.58       7.10          0.48   novodra   
73       mette   sandgreen         7.56       7.08          0.48   auralin   
217     taylah    mobourne         7.90       7.51          0.39   auralin   
478      david  gustafsson         7.72       7.28          0.44   novodra   
622     steven         roy         7.87       7.43          0.44   novodra   

    start_dose end_dose  
196       51u       60u  
575       43u       48u  
73        36u       46u  
217       41u       50u  
478       33u       34u  
622       41u       43u

#### Adverse reaction should be part of the `treatments` table

##### Define
Merge the adverse_reaction column to the treatments table, joining on given name and surname.

##### Code

In [39]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name','surname'], how='left')

##### Test

In [40]:
# Your testing code here
treatments_clean.sample(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
98       peter    pospíšil         7.76       7.34          0.42   auralin   
326   miłosław  wiśniewski         7.51       7.08          0.43   novodra   
286     torben   mortensen         7.80       7.40          0.40   novodra   

    start_dose end_dose           adverse_reaction  
98        51u       60u                        NaN  
326       34u       33u  injection site discomfort  
286       44u       40u                        NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the patients table, then convert these names to lower case to join with treatments. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore)

##### Code

In [41]:
# Your cleaning code here

# Isolate the patient ID and names in the patients table,
names = patients_clean[['patient_id', 'given_name', 'surname']]

# Then convert these names to lower case 
names.given_name = names.given_name.str.lower()
names.surname = names.surname.str.lower()
# to join with treatments
treatments_clean = pd.merge(treatments_clean, names, on=['given_name', 'surname'])

# Then drop the given name and surname columns 
treatments_clean.drop(columns=['given_name', 'surname'], axis=1, inplace=True)

C:\Users\Mahmoud\Anaconda3\lib\site-packages\pandas\core\generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


##### Test

In [42]:
# Confirm the merge was executed correctly
treatments_clean.sample(4)

hba1c_start  hba1c_end  hba1c_change treatment start_dose end_dose  \
138         7.85       7.45          0.40   auralin       57u       67u   
238         7.94       7.57          0.37   novodra       32u       27u   
322         7.73       7.34          0.39   novodra       20u       21u   
22          7.67       7.37          0.30   auralin       43u       47u   

    adverse_reaction  patient_id  
138              NaN         107  
238              NaN         219  
322              NaN          49  
22               NaN         146

In [43]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert the zip code column's data type from a float to a string using astype, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

##### Code

In [44]:
# Your cleaning code here
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, side='left', fillchar='0')

# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [45]:
# Your testing code here
patients_clean[patients_clean.zip_code == '0000n']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [46]:
patients_clean.sample(4)

patient_id assigned_sex given_name      surname               address  \
320         321         male   Noriyuki        Sakai       353 Whaley Lane   
325         326       female       Sara  Karjalainen  720 Tator Patch Road   
262         263       female    Rebecca       Priest      2020 Gore Street   
115         116       female   Caroline         Bois      1350 Meadow Lane   

          city       state zip_code        country   birthdate  weight  \
320  Milwaukee          WI    53202  United States   4/29/1934   150.3   
325    Chicago    Illinois    60601  United States   8/20/1946   172.5   
262    Houston          TX    77092  United States   3/26/1988   194.0   
115   Lakeport  California    95453  United States  12/26/1985   151.4   

     height   bmi  phone_number                        email  
320      70  21.6  262-565-9052      NoriyukiSakai@gustr.com  
325      63  30.6  312 856 6826  SaraKarjalainen@armyspy.com  
262      68  29.5  713-556-4328    RebeccaRPriest@dayrep.com  
115      64  26.0  707 262 6503       CarolineBois@rhyta.com

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [47]:
# Your cleaning code here
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [48]:
# Your testing code here
patients_clean[patients_clean.given_name == 'Tim']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [49]:
# Your cleaning code here

# states to abbreviate
state_abbreviations = {'California':'CA', 'New York':'NY', 'Illinois':'IL', 'Florida':'FL', 'Nebraska':'NE'}

def abbreviate_state(patient):
    if patient['state'] in state_abbreviations.keys():
        return state_abbreviations[ patient['state'] ]
    else:
        return patient['state']

# Apply a function that converts full state name to state abbreviation
patients_clean.state = patients_clean.apply(abbreviate_state, axis=1)

##### Test

In [50]:
# Your testing code here
patients.state.value_counts()

California    36
TX            32
New York      25
CA            24
MA            22
NY            22
PA            18
GA            15
Illinois      14
OH            14
LA            13
MI            13
Florida       13
OK            13
NJ            12
VA            11
WI            10
IL            10
MS            10
MN             9
AL             9
TN             9
FL             9
IN             9
NC             8
KY             8
WA             8
MO             7
NV             6
KS             6
ID             6
SC             5
IA             5
CT             5
AZ             4
ME             4
RI             4
Nebraska       4
CO             4
ND             4
AR             4
WV             3
SD             3
MD             3
OR             3
DE             3
DC             2
VT             2
NE             2
MT             2
NM             1
AK             1
WY             1
NH             1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
Replace given name for rows in the patients table that have a given name of 'Dsvid' with 'David'.

##### Code

In [51]:
patients_clean[patients_clean.given_name=='Dsvid']

patient_id assigned_sex given_name     surname             address  \
8           9         male      Dsvid  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

In [52]:
# Your cleaning code here
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [53]:
# Your testing code here
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
Convert `assigned_sex` and `state` to categorical data types. `Zip_code` data type was already addressed above. Convert `birthdate` to datetime data type. Strip the letter 'u' in `start_dose` and `end_dose` and convert those columns to data type integer.

##### Code

In [54]:
# Your cleaning code here

# Convert assigned_sex and state to categorical data types. 
patients_clean.assigned_sex = patients_clean.assigned_sex.astype(str)
patients_clean.state = patients_clean.state.astype(str)

# Convert birthdate to datetime data type.
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip the letter 'u' in start_dose and end_dose and convert those columns to data type integer.
treatments_clean.start_dose = treatments_clean.start_dose.str[:-2].astype(int)
treatments_clean.end_dose = treatments_clean.end_dose.str[:-2].astype(int)

##### Test

In [55]:
# Your testing code here
patients_clean.dtypes

patient_id               int64
assigned_sex            object
given_name              object
surname                 object
address                 object
city                    object
state                   object
zip_code                object
country                 object
birthdate       datetime64[ns]
weight                 float64
height                   int64
bmi                    float64
phone_number            object
email                   object
dtype: object

In [56]:
treatments_clean.dtypes

hba1c_start         float64
hba1c_end           float64
hba1c_change        float64
treatment            object
start_dose            int32
end_dose              int32
adverse_reaction     object
patient_id            int64
dtype: object

#### Multiple phone number formats

##### Define
Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [57]:
# Your cleaning code here
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'(\D+)', '').str.pad(11, side='left', fillchar='1')

##### Test

In [58]:
# Your testing code here
patients_clean.phone_number.sample(5)

264            NaN
391    17869704206
307    12282372271
200    18566555415
320    12625659052
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
Remove the non-recoverable John Doe records from the `patients` table.


##### Code

In [59]:
patients_clean[patients_clean.surname == 'Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   
229    NY    12345  United States 1975-01-01   180.0      72  24.4   
237    NY    12345  United States 1975-01-01   180.0      72  24.4   
244    NY    12345  United States 1975-01-01   180.0      72  24.4   
251    NY    12345  United States 1975-01-01   180.0      72  24.4   
277    NY    12345  United States 1975-01-01   180.0      72  24.4   

    phone_number              email  
215  11234567890  johndoe@email.com  
229  11234567890  johndoe@email.com  
237  11234567890  johndoe@email.com  
244  11234567890  johndoe@email.com  
251  11234567890  johndoe@email.com  
277  11234567890  johndoe@email.com

In [60]:
# Your cleaning code here
patients_clean.drop([229, 237, 244, 251, 277], axis=0, inplace=True)

##### Test

In [61]:
# Your testing code here
patients_clean[patients_clean.surname == 'Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   

    phone_number              email  
215  11234567890  johndoe@email.com

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the patients table. These are the nicknames, which happen to also not be in the treatments table (removing the wrong name would create a consistency issue between the patients and treatments table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [62]:
patients_clean.surname.value_counts()

Jakobsen      3
Taylor        3
Aranda        2
Gersten       2
Ogochukwu     2
             ..
Horvat        1
Fiorentino    1
Totth         1
Cotton        1
Knutsen       1
Name: surname, Length: 466, dtype: int64

In [63]:
# Your cleaning code here
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [64]:
# Your testing code here
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [65]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [66]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

#### kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [70]:
patients_clean[patients_clean.surname == 'Zaitseva'].weight

210    48.8
Name: weight, dtype: float64

In [71]:
# Your cleaning code here
patients_clean.loc[patients_clean.surname == 'Zaitseva', 'weight'] = 48.8*2.20462

##### Test

In [72]:
# Your testing code here
patients_clean.weight.sort_values()

459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 495, dtype: float64